pilot/pkg/dns/dns.go

```diff
@@ -0,0 +1,344 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dns
+
+import (
+	"net"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/miekg/dns"
+
+	nds "istio.io/istio/pilot/pkg/proto"
+	"istio.io/pkg/log"
+)
+
+// Holds configurations for the DNS downstreamServer in Istio Agent
+type LocalDNSServer struct {
+	// Holds the pointer to the DNS lookup table
+	lookupTable atomic.Value
+
+	downstreamMux    *dns.ServeMux
+	downstreamServer *dns.Server
+
+	// This is the upstreamClient used to make upstream DNS queries
+	// in case the data is not in our cache.
+	upstreamClient    *dns.Client
+	resolvConfServers []string
+	searchNamespaces  []string
+	// The namespace where the proxy resides
+	// determines the hosts used for shortname resolution
+	proxyNamespace string
+	// Optimizations to save space and time
+	proxyDomain      string
+	proxyDomainParts []string
+}
+
+// Borrowed from https://github.com/coredns/coredns/blob/master/plugin/hosts/hostsfile.go
+type LookupTable struct {
+	// The key is a FQDN matching a DNS query (like example.com.), the value is pre-created DNS RR records
+	// of A or AAAA type as appropriate.
+	name4 map[string][]dns.RR
+	name6 map[string][]dns.RR
+	cname map[string][]dns.RR
+}
+
+const (
+	// In case the client decides to honor the TTL, keep it low so that we can always serve
+	// the latest IP for a host.
+	defaultTTL = 30
```

Comment:
is this used only for our direct responses? forwarded requests will still respect the TTL right?

Reply:
CoreDNS allows configuring this, may be worth making this an option

Reply:
Yes thats on my todo list.